File: ncore/utils/db_tool_v2/libs/sqlite.py
import sqlite3
from typing import Dict, List, Union, Tuple, Optional, Any
from pycore.globalvers import appenv, appdir
from pycore.utils_linux import file
from pycore.dbmode.baseclass.dbtoolbase import DBToolBase

class Sqlite(DBToolBase):

    def __init__(self, appenv_or_dburl: str, config_file: Dict=None,debug:bool=False):
        db_url = self.get_db_url_from_config(appenv_or_dburl)
        self.db_url = db_url
        self.config_file = config_file
        self.include_filter = []
        self.exclude_filter = []
        self.show_sql = debug

    def set_debug(self,debug):
        self.show_sql = debug

    def get_db_url_from_config(self, appenv_or_dburl):
        return super().get_db_url_from_config_sqlite(appenv_or_dburl)

    def set_include_filter(self, include_filter: List[str]):
        self.include_filter = include_filter

    def set_exclude_filter(self, exclude_filter: List[str]):
        self.exclude_filter = exclude_filter

    def filter_tables(self, tables: List[str]) -> List[str]:
        system_tables = {'sqlite_sequence', 'sqlite_stat1', 'sqlite_stat2', 'sqlite_stat3', 'sqlite_stat4'}
        filtered_tables = [table for table in tables if table not in system_tables]

        if self.include_filter:
            filtered_tables = [table for table in filtered_tables if table in self.include_filter]
        if self.exclude_filter:
            filtered_tables = [table for table in filtered_tables if table not in self.exclude_filter]
        return filtered_tables

    def config_file_exists(self) -> bool:
        try:
            with open(self.config_file, 'r') as f:
                return True
        except FileNotFoundError:
            return False

    def _connect(self):
        self.conn = sqlite3.connect(self.db_url)
        self.cursor = self.conn.cursor()

    def _close(self):
        if self.cursor:
            self.cursor.close()
        if self.conn:
            self.conn.close()
        self.cursor = None
        self.conn = None

    def _execute(self, sql: str, params: Union[Tuple, List[Tuple]] = (), many: bool = False):
        self._connect()

        if self.show_sql:
            print(sql)
        if many:
            self.cursor.executemany(sql, params)
        else:
            self.cursor.execute(sql, params)
        self.conn.commit()
        return self.cursor


    def init_database(self, table_maps: Dict[str, Dict[str, Dict[str, Union[str, bool, Optional[str]]]]]):
        self.create_tables(table_maps)

    def insert_one(self, tabname: str, data: Dict, result_id: bool = True):
        keys = ', '.join(data.keys())
        question_marks = ', '.join(['?' for _ in data])
        values = tuple(data.values())
        sql = f"INSERT INTO {tabname} ({keys}) VALUES ({question_marks})"
        cursor = self._execute(sql, values)
        self._close()
        if result_id:
            return cursor.lastrowid

    def update_one(self, tabname: str, data: Dict, conditions: Dict):
        set_clause = ', '.join([f"{key} = ?" for key in data.keys()])
        where_clause, values = self._build_conditions(conditions)
        values = tuple(data.values()) + values
        sql = f"UPDATE {tabname} SET {set_clause} WHERE {where_clause}"
        self._execute(sql, values)
        self._close()

    def read_one(self, tablename: str, conditions: Dict = None, select: str = "*") -> Optional[Dict[str, Any]]:
        where_clause, values = self._build_conditions(conditions)
        sql = f"SELECT {select} FROM {tablename} WHERE {where_clause}" if where_clause else f"SELECT {select} FROM {tablename}"
        cursor = self._execute(sql, values)
        read_result = cursor.fetchone()
        result = {}
        if read_result:
            read_result = [col[0] for col in cursor.description]
            result = dict(zip(columns, read_result))
        self._close()
        return result

    def insert_many(self, tabname: str, data: List[Dict], result_id: bool = True):
        if not data:
            return
        keys = ', '.join(data[0].keys())
        question_marks = ', '.join(['?' for _ in data[0]])
        values = [tuple(d.values()) for d in data]
        sql = f"INSERT INTO {tabname} ({keys}) VALUES ({question_marks})"
        cursor = self._execute(sql, values, many=True)
        result_id = cursor.lastrowid
        self._close()
        if result_id:
            return result_id

    def update_many(self, tabname: str, data: List[Dict], conditions: Dict):
        if not data:
            return
        set_clause = ', '.join([f"{key} = ?" for key in data[0].keys()])
        where_clause, cond_values = self._build_conditions(conditions)
        values = [tuple(d.values()) + cond_values for d in data]
        sql = f"UPDATE {tabname} SET {set_clause} WHERE {where_clause}"
        self._execute(sql, values, many=True)
        self._close()

    def read_many(self, tablename: str, conditions: Dict = None, limit: Tuple[int, int] = (0, 1000),
                  select: str = "*", sort: Dict[str, str] = None, print_sql: bool = False) -> List[Dict[str, Any]]:
        where_clause, values = self._build_conditions(conditions)
        limit_clause = f" LIMIT {limit[0]}, {limit[1]}"
        sort_clause = ' ORDER BY ' + ', '.join([f"{key} {value}" for key, value in sort.items()]) if sort else ''
        sql = f"SELECT {select} FROM {tablename} WHERE {where_clause}{sort_clause}{limit_clause}" if where_clause else f"SELECT {select} FROM {tablename}{sort_clause}{limit_clause}"
        cursor = self._execute(sql, values)
        result = cursor.fetchall()
        columns = [col[0] for col in cursor.description]
        self._close()
        data = [dict(zip(columns, row)) for row in result]
        return data

    def delete(self, tabname: str, conditions: Dict = None, physical: bool = False):
        where_clause, values = self._build_conditions(conditions)
        if physical:
            sql = f"DELETE FROM {tabname} WHERE {where_clause}"
        else:
            sql = f"UPDATE {tabname} SET deleted = 1 WHERE {where_clause}"
        self._execute(sql, values)
        self._close()

    def filter_query(self, query_obj, table_class, conditions: Dict = None):
        pass

    def get_session(self):
        self._connect()
        return self.conn

    def get_table(self, tabname: str) -> Dict[str, Dict[str, Union[int, str, bool, Optional[str]]]]:
        sql = f"PRAGMA table_xinfo({tabname})"
        cursor = self._execute(sql)
        columns = cursor.fetchall()
        self._close()

        result = {
            col[1]: {
                "cid": col[0],
                "name": col[1],
                "type": col[2],
                "notnull": bool(col[3]),
                "default_value": col[4],
                "primary_key": bool(col[5]),
                "hidden": bool(col[6])
            } for col in columns
        }
        return result

    def get_tablemaps(self) -> Dict[str, Dict[str, Dict[str, Union[str, bool, Optional[str]]]]]:
        sql = "SELECT name FROM sqlite_master WHERE type='table'"
        cursor = self._execute(sql)
        tables = [row[0] for row in cursor.fetchall()]
        tables = self.filter_tables(tables)
        table_maps = {table: self.get_table(table) for table in tables}
        self._close()
        return table_maps

    def show_tablemap(self):
        table_maps = self.get_tablemaps()
        for table, fields in table_maps.items():
            print(f"Table: {table}")
            for field, info in fields.items():
                print(f"  Field: {field} - {info}")

    def create_table(self, tablename: str, fields: Dict[str, Dict[str, Union[int, str, bool, Optional[str]]]]):
        if self.table_exists(tablename):
            self.info(f"Table '{tablename}' already exists. Skipping creation.")
            return

        columns = []
        for name, info in fields.items():
            column_def = f"{name} {info['type']}"
            if info.get('primary_key'):
                column_def += " PRIMARY KEY AUTOINCREMENT"
            if info.get('notnull'):
                column_def += " NOT NULL"
            if info.get('default_value') is not None:
                column_def += f" DEFAULT {info['default_value']}"
            columns.append(column_def)
        columns_str = ", ".join(columns)
        sql = f"CREATE TABLE IF NOT EXISTS {tablename} ({columns_str})"
        self._execute(sql)
        self._close()
        self.success(f"Table '{tablename}' created successfully.")

    def create_tables(self, table_maps: Dict[str, Dict[str, Dict[str, Union[int, str, bool, Optional[str]]]]]):
        for tablename, fields in table_maps.items():
            self.create_table(tablename, fields)


    def drop_table(self, tablename: str):
        sql = f"DROP TABLE IF EXISTS {tablename}"
        self._execute(sql)
        self._close()

    def table_exists(self, tablename: str) -> bool:
        sql = f"SELECT name FROM sqlite_master WHERE type='table' AND name='{tablename}'"
        cursor = self._execute(sql)
        is_table = bool(cursor.fetchone())
        self._close()
        return is_table

    def migrate_data(self, origin_db,target_db):
        super().migrate_data(origin_db,target_db)
        pass

    def _build_conditions(self, conditions: Dict[str, str]) -> Tuple[str, Tuple]:
        return super().build_conditions_sqlite(conditions=conditions)

File: ncore/utils/db_tool/libs/sqlite.js
const sqlite3 = require('sqlite3');
    const { open } = require('sqlite');
    const sqliteTool = require('../utils/sqlite_tool.js');
    const Base = require('#@base');

    class SQLite extends Base {
      constructor(configOrDbFile, initConfigFile = '', debug = false) {
        super()
        this.dbFile = typeof configOrDbFile === 'string' ? configOrDbFile : sqliteTool.getDbFileFromConfig(configOrDbFile);
        this.configFile = initConfigFile;
        this.includeFilter = [];
        this.excludeFilter = [];
        this.showSql = debug;
        this.connection = null;
      }

      async connect() {
        if (!this.connection) {
          this.connection = await open({
            filename: this.dbFile,
            driver: sqlite3.Database
          });
        }
        return this.connection;
      }

      async close() {
        if (this.connection) {
          await this.connection.close();
          this.connection = null;
        }
      }

      setDebug(debug) {
        this.showSql = debug;
      }

      setIncludeFilter(includeFilter) {
        this.includeFilter = includeFilter;
      }

      setExcludeFilter(excludeFilter) {
        this.excludeFilter = excludeFilter;
      }

      async execute(sql, params = []) {
        const connection = await this.connect();
        if (this.showSql) {
          console.log(sql);
        }
        const result = await connection.run(sql, params);
        return result;
      }

      async query(sql, params = []) {
        const connection = await this.connect();
        if (this.showSql) {
          console.log(sql);
        }
        const result = await connection.all(sql, params);
        return result;
      }

      async initDatabase(tableMaps) {
        await this.createTables(tableMaps);
      }

      async insertOne(tabname, data) {
        const keys = Object.keys(data).join(', ');
        const placeholders = Object.keys(data).map(() => '?').join(', ');
        const sql = `INSERT INTO ${tabname} (${keys}) VALUES (${placeholders})`;
        const result = await this.execute(sql, Object.values(data));
        return result.lastID;
      }

      async updateOne(tabname, data, conditions) {
        const setClause = Object.keys(data).map(key => `${key} = ?`).join(', ');
        const [whereClause, values] = sqliteTool.buildConditions(conditions);
        const sql = `UPDATE ${tabname} SET ${setClause} WHERE ${whereClause}`;
        await this.execute(sql, [...Object.values(data), ...values]);
      }

      async readOne(tablename, conditions = {}, select = '*') {
        const [whereClause, values] = sqliteTool.buildConditions(conditions);
        const sql = `SELECT ${select} FROM ${tablename} ${whereClause ? `WHERE ${whereClause}` : ''}`;
        const result = await this.query(sql, values);
        return result[0] || null;
      }

      async insertMany(tabname, data) {
        if (data.length === 0) return;

        const keys = Object.keys(data[0]).join(', ');
        const placeholders = Object.keys(data[0]).map(() => '?').join(', ');
        const sql = `INSERT INTO ${tabname} (${keys}) VALUES (${placeholders})`;

        const values = data.map(row => Object.values(row));
        const connection = await this.connect();
        const stmt = await connection.prepare(sql);
        try {
          for (const row of values) {
            await stmt.run(row);
          }
        } finally {
          await stmt.finalize();
        }
      }

      async updateMany(tabname, data, conditions) {
        if (data.length === 0) return;

        const setClause = Object.keys(data[0]).map(key => `${key} = ?`).join(', ');
        const [whereClause, condValues] = sqliteTool.buildConditions(conditions);
        const sql = `UPDATE ${tabname} SET ${setClause} WHERE ${whereClause}`;
        
        const connection = await this.connect();
        const stmt = await connection.prepare(sql);
        try {
          for (const row of data) {
            await stmt.run([...Object.values(row), ...condValues]);
          }
        } finally {
          await stmt.finalize();
        }
      }

      async readMany(tablename, conditions = {}, limit = [0, 1000], select = '*', sort = null) {
        const [whereClause, values] = sqliteTool.buildConditions(conditions);
        const sortClause = sort ? ` ORDER BY ${Object.entries(sort).map(([k, v]) => `${k} ${v}`).join(', ')}` : '';
        const limitClause = ` LIMIT ${limit[0]}, ${limit[1]}`;
        const sql = `SELECT ${select} FROM ${tablename} ${whereClause ? `WHERE ${whereClause}` : ''} ${sortClause} ${limitClause}`;
        return await this.query(sql, values);
      }

      async delete(tabname, conditions = {}, physical = false) {
        const [whereClause, values] = sqliteTool.buildConditions(conditions);
        const sql = physical ? `DELETE FROM ${tabname} WHERE ${whereClause}` : `UPDATE ${tabname} SET deleted = 1 WHERE ${whereClause}`;
        await this.execute(sql, values);
      }

      async getTable(tabname) {
        const sql = `PRAGMA table_info(${tabname})`;
        const columns = await this.query(sql);
        return sqliteTool.reduceColumns(columns);
      }

      async getTableMaps() {
        const tables = await this.query("SELECT name FROM sqlite_master WHERE type='table'");
        const tableNames = tables.map(row => row.name);
        const filteredTables = sqliteTool.filterTables(tableNames, this.includeFilter, this.excludeFilter);
        const tableMaps = {};
        for (const table of filteredTables) {
          tableMaps[table] = await this.getTable(table);
        }
        return tableMaps;
      }

      async showTableMap() {
        const tableMaps = await this.getTableMaps();
        for (const [table, fields] of Object.entries(tableMaps)) {
          console.log(`\n\tTable: ${table}`);
          for (const [field, info] of Object.entries(fields)) {
            console.log(`\t\t ${field} : ${JSON.stringify(info)}`);
          }
        }
      }

      async createTable(tablename, fields) {
        if (await this.tableExists(tablename)) {
          console.log(`Table '${tablename}' already exists. Skipping creation.`);
          return;
        }

        const columns = Object.entries(fields).map(([name, info]) => {
          let columnDef = `${name} ${sqliteTool.getColumnType(info.type)}`;
          if (info.primary_key) columnDef += ' PRIMARY KEY';
          if (info.notnull) columnDef += ' NOT NULL';
          if (info.default_value !== undefined) columnDef += ` DEFAULT '${info.default_value}'`;
          return columnDef;
        });

        const sql = `CREATE TABLE ${tablename} (${columns.join(', ')})`;
        await this.execute(sql);
        console.log(`Table '${tablename}' created successfully.`);
      }

      async createTables(tableMaps) {
        for (const [tablename, fields] of Object.entries(tableMaps)) {
          await this.createTable(tablename, fields);
        }
      }

      async dropTable(tablename) {
        const sql = `DROP TABLE IF EXISTS ${tablename}`;
        await this.execute(sql);
      }

      async tableExists(tablename) {
        const sql = `SELECT name FROM sqlite_master WHERE type='table' AND name=?`;
        const result = await this.query(sql, [tablename]);
        return result.length > 0;
      }

      async migrateData(targetDb) {
        // Implement your data migration logic here
      }
    }

    module.exports = SQLite;